test/integration/controller/jobset/jobset_controller_test.go

```diff
@@ -0,0 +1,667 @@
+		}, util.Timeout, util.Interval).Should(gomega.BeTrue())
+		gomega.Expect(createdWorkload.Status.Admission).Should(gomega.BeNil())
+
+		ginkgo.By("checking the JobSet is unsuspended and selectors added when workload is assigned again")
+		admission = testing.MakeAdmission(clusterQueue.Name).
+			PodSets(
+				kueue.PodSetAssignment{
+					Name: "replicated-job-1",
+					Flavors: map[corev1.ResourceName]kueue.ResourceFlavorReference{
+						corev1.ResourceCPU: "on-demand",
+					},
+					Count: pointer.Int32(createdWorkload.Spec.PodSets[0].Count),
+				},
+				kueue.PodSetAssignment{
+					Name: "replicated-job-2",
+					Flavors: map[corev1.ResourceName]kueue.ResourceFlavorReference{
+						corev1.ResourceCPU: "spot",
+					},
+					Count: pointer.Int32(createdWorkload.Spec.PodSets[1].Count),
+				},
+			).
+			Obj()
+		gomega.Expect(util.SetAdmission(ctx, k8sClient, createdWorkload, admission)).Should(gomega.Succeed())
+		gomega.Eventually(func() bool {
+			if err := k8sClient.Get(ctx, lookupKey, createdJobSet); err != nil {
+				return false
+			}
+			return !*createdJobSet.Spec.Suspend
+		}, util.Timeout, util.Interval).Should(gomega.BeTrue())
+
+		gomega.Expect(len(createdJobSet.Spec.ReplicatedJobs[0].Template.Spec.Template.Spec.NodeSelector)).Should(gomega.Equal(1))
+		gomega.Expect(createdJobSet.Spec.ReplicatedJobs[0].Template.Spec.Template.Spec.NodeSelector[labelKey]).Should(gomega.Equal(onDemandFlavor.Name))
+		gomega.Expect(len(createdJobSet.Spec.ReplicatedJobs[1].Template.Spec.Template.Spec.NodeSelector)).Should(gomega.Equal(1))
+		gomega.Expect(createdJobSet.Spec.ReplicatedJobs[1].Template.Spec.Template.Spec.NodeSelector[labelKey]).Should(gomega.Equal(spotFlavor.Name))
+
+		gomega.Consistently(func() bool {
+			if err := k8sClient.Get(ctx, wlLookupKey, createdWorkload); err != nil {
+				return false
+			}
+			return workload.IsAdmitted(createdWorkload)
+		}, util.ConsistentDuration, util.Interval).Should(gomega.BeTrue())
+
+		ginkgo.By("checking the workload is finished when JobSet is completed")
+		createdJobSet.Status.Conditions = append(createdJobSet.Status.Conditions,
+			metav1.Condition{
+				Type:               string(jobsetapi.JobSetCompleted),
+				Status:             metav1.ConditionStatus(corev1.ConditionTrue),
+				Reason:             "AllJobsCompleted",
+				Message:            "jobset completed successfully",
+				LastTransitionTime: metav1.Now(),
+			})
+		gomega.Expect(k8sClient.Status().Update(ctx, createdJobSet)).Should(gomega.Succeed())
+		gomega.Eventually(func() bool {
+			err := k8sClient.Get(ctx, wlLookupKey, createdWorkload)
+			if err != nil {
+				return false
+			}
+			return apimeta.IsStatusConditionTrue(createdWorkload.Status.Conditions, kueue.WorkloadFinished)
+		}, util.Timeout, util.Interval).Should(gomega.BeTrue())
+	})
+})
+
+var _ = ginkgo.Describe("JobSet controller for workloads when only jobs with queue are managed", ginkgo.Ordered, ginkgo.ContinueOnFailure, func() {
```

Comment:
Can we test this at unit level somehow? We already have coverage for this for Job and this is just using the same reconciler. So I think we can remove this test for the sake of a leaner test suite.

Reply:
maybe if we add a testFake GenericJob to be used in jobframework unittest. that could be useful since the jobframework/reconciler.go is probably the most used and is only 5% covered by unit test. #922